### PR TITLE
Fixed #28300 -- Allowed GDALRaster objects to use the vsimem filesystem.

### PR DESCRIPTION
**Pull request by yellowcap**
https://code.djangoproject.com/ticket/28300

### CODE REVIEW

**Comment by timgraham** [comment]
buildbot, test on xenial.

File: tests/gis_tests/gdal_tests/test_raster.py [hunk 1]
```diff
@@ -121,7 +121,7 @@ def test_memory_based_raster_creation(self):
         # Assert data is same as original input
         self.assertEqual(result, list(range(256)))
 
-    def test_file_based_raster_creation(self):
+    def test_file_path_based_raster_creation(self):
```
Comment (timgraham) [inline tests/gis_tests/gdal_tests/test_raster.py]: Should this rename be in a separate commit?

File: docs/releases/2.0.txt [hunk 2]
```diff
@@ -86,6 +86,10 @@ Minor features
 * Allowed passing driver-specific creation options to
   :class:`~django.contrib.gis.gdal.GDALRaster` objects using ``papsz_options``.
 
+* Allowed creating :class:`~django.contrib.gis.gdal.GDALRaster` objects in
+  GDAL's internal virtual filesystem. Rasters can now be created from and
+  converted to binary data in memory directly.
```
Comment (timgraham) [inline docs/releases/2.0.txt]: Link to the new documentation in the release note.

File: docs/ref/contrib/gis/gdal.txt [hunk 3]
```diff
@@ -1731,6 +1770,55 @@ Key              Default                           Usage
...
+    # Create a new GTiff raster in a view and return it as a file.
+    >>> rst = GDALRaster({
+    ...     "name": "/vsimem/temporarymemfile",
+    ...     "driver": "tif",
+    ...     "width": 6, "height": 6, "srid": 3086,
+    ...     "origin": [500000, 400000],
+    ...     "scale": [100, -100],
+    ...     "bands": [{"data": range(36), "nodata_value": 99}]
+    ... })
+    >>> return HttpResponse(rast.vsi_buffer, 'image/tiff')
```
Comment (timgraham) [inline docs/ref/contrib/gis/gdal.txt]: A return statement doesn't make sense in a shell session. Also, HttpResponse import is missing.

File: docs/ref/contrib/gis/gdal.txt [hunk 4]
```diff
@@ -1731,6 +1770,55 @@ Key              Default                           Usage
...
+dictionary representation, and provide a ``name`` argument that starts with
+``/vsimem/`` (for detail on the dictionary representation see
+:ref:`gdal-raster-ds-input`). For virtual file based rasters, the
+:attr:`~GDALRaster.vsi_buffer` will return the ``bytes`` representation of the
+raster. 
+::
+
+    # Create a new GTiff raster in a view and return it as a file.
+    >>> rst = GDALRaster({
+    ...     "name": "/vsimem/temporarymemfile",
```
Comment (timgraham) [inline docs/ref/contrib/gis/gdal.txt]: Use single quotes

File: docs/ref/contrib/gis/gdal.txt [hunk 5]
```diff
@@ -1731,6 +1770,55 @@ Key              Default                           Usage
...
+:class:`GDALRaster` objects are created in the virtual filesystem when a
+``bytes`` object is provided as input, or when the file path starts with
+``/vsimem/``.
+
+Input provided as ``bytes`` has to be a full binary representation of a file.
+For instance::
+
+    # Read a raster as file object from a remote source.
+    >>> import urllib
+    >>> dat = urllib.urlopen('http://example.com/raster.tif').read()
```
Comment (timgraham) [inline docs/ref/contrib/gis/gdal.txt]: urllib.urlopen looks like Python 2

File: docs/ref/contrib/gis/gdal.txt [hunk 6]
```diff
@@ -1105,15 +1105,25 @@ blue.
...
-    can be created from different input sources (using the sample data from the
-    GeoDjango tests, see also the :ref:`gdal_sample_data` section). For a
-    detailed description of how to create rasters using dictionary input, see
-    the :ref:`gdal-raster-ds-input` section.
+    defines the raster source, the second parameter defines if a raster should be
+    opened in write mode. For newly created rasters the second parameter is ignored
+    and the new raster is always created in write mode.
+
+    The first parameter can take three forms: a string representing a file path, a
+    dictionary with values defining a new raster, or a bytes object representing a raster
```
Comment (timgraham) [inline docs/ref/contrib/gis/gdal.txt]: Line length is exceed 79 characters in the new text.

File: docs/ref/contrib/gis/gdal.txt [hunk 7]
```diff
@@ -1731,6 +1770,55 @@ Key              Default                           Usage
...
+    >>> rst = GDALRaster({
+    ...     "name": "/vsimem/temporarymemfile",
+    ...     "driver": "tif",
+    ...     "width": 6, "height": 6, "srid": 3086,
+    ...     "origin": [500000, 400000],
+    ...     "scale": [100, -100],
+    ...     "bands": [{"data": range(36), "nodata_value": 99}]
+    ... })
+    >>> return HttpResponse(rast.vsi_buffer, 'image/tiff')
+
```
Comment (timgraham) [inline docs/ref/contrib/gis/gdal.txt]: 1 blank only

File: docs/ref/contrib/gis/gdal.txt [hunk 8]
```diff
@@ -1731,6 +1770,55 @@ Key              Default                           Usage
...
+    # Read a raster as file object from a remote source.
+    >>> import urllib
+    >>> dat = urllib.urlopen('http://example.com/raster.tif').read()
+    # Instantiate a raster from the bytes object.
+    >>> rst = GDALRaster(dat)
+    # The name starts with /vsimem/, indicating that the raster lives in the
+    # virtual filesystem.
+    >>> rst.name
+    /vsimem/da300bdb-129d-49a8-b336-e410a9428dad
```
Comment (timgraham) [inline docs/ref/contrib/gis/gdal.txt]: Should that be a quoted string?

File: docs/ref/contrib/gis/gdal.txt [hunk 9]
```diff
@@ -1731,6 +1770,55 @@ Key              Default                           Usage
...
+
+To create a new virtual file based raster from scratch, use the ``ds_input``
+dictionary representation, and provide a ``name`` argument that starts with
+``/vsimem/`` (for detail on the dictionary representation see
+:ref:`gdal-raster-ds-input`). For virtual file based rasters, the
+:attr:`~GDALRaster.vsi_buffer` will return the ``bytes`` representation of the
```
Comment (timgraham) [inline docs/ref/contrib/gis/gdal.txt]: will return -> returns

File: django/contrib/gis/gdal/raster/const.py [hunk 10]
```diff
@@ -43,3 +43,13 @@
     'Average': 5,
     'Mode': 6,
 }
+
+# Fixed basepath for buffer based GDAL in memory files.
+VSI_FILESYSTEM_BASEPATH = '/vsimem/'
+
+# Flag for managing ownership over buffer when creating VSI in
+# memory files.
+VSI_TAKE_BUFFER_OWNERSHIP = False
```
Comment (timgraham) [inline django/contrib/gis/gdal/raster/const.py]: Are these boolean constants just for readability or are these settings meant to be monkeypatched for alternate behavior?

File: django/contrib/gis/gdal/raster/source.py [hunk 11]
```diff
@@ -66,13 +73,39 @@ def __init__(self, ds_input, write=False):
...
                 # GDALOpen will auto-detect the data source type.
                 self._ptr = capi.open_ds(force_bytes(ds_input), self._write)
             except GDALException as err:
                 raise GDALException('Could not open the datasource at "{}" ({}).'.format(ds_input, err))
+        elif isinstance(ds_input, bytes):
+            # A new raster needs to be created in write mode.
+            self._write = 1
+            # Get size of buffer.
+            size = sys.getsizeof(ds_input)
+            # Pass data to ctypes, keep reference to the ctypes object.
```
Comment (timgraham) [inline django/contrib/gis/gdal/raster/source.py]: , keeping a reference to the cyptes object so that the vsimem file...

File: django/contrib/gis/gdal/raster/source.py [hunk 12]
```diff
@@ -151,6 +184,13 @@ def __init__(self, ds_input, write=False):
         else:
             raise GDALException('Invalid data source input type: "{}".'.format(type(ds_input)))
 
+    def __del__(self):
```
Comment (timgraham) [inline django/contrib/gis/gdal/raster/source.py]: A task in https://code.djangoproject.com/ticket/23919 is "Evaluate replacement of custom `__del__` methods by  [weakref.finalize()](https://docs.python.org/3/library/weakref.html#weakref.finalize)". I wonder if you have any input about that?

File: django/contrib/gis/gdal/raster/const.py (hunk 10 above)
Comment (yellowcap) [inline django/contrib/gis/gdal/raster/const.py]: These are meant for readability only.

File: django/contrib/gis/gdal/raster/source.py (hunk 12 above)
Comment (yellowcap) [inline django/contrib/gis/gdal/raster/source.py]: Unfortunately I can't give input on that. I have never heard of `weakref` before, and I don't know enough about pythons gc etc to have an informed opinion.

File: django/contrib/gis/gdal/raster/source.py [hunk 13]
```diff
@@ -66,13 +73,39 @@ def __init__(self, ds_input, write=False):
 
         # If input is a valid file path, try setting file as source.
         if isinstance(ds_input, str):
-            if not os.path.exists(ds_input):
-                raise GDALException('Unable to read raster source input "{}"'.format(ds_input))
+            if not ds_input.startswith(VSI_FILESYSTEM_BASEPATH) and not os.path.exists(ds_input):
+                raise GDALException('Opening rasters from vsi file paths is not allowed.')
```
Comment (yellowcap) [inline django/contrib/gis/gdal/raster/source.py]: Thanks for spotting this, the error message like this does not make sense. I had a previous version where direct opening of vsi paths was not allowed. But it should work like files, its the user's responsibility when opening the same vsi file twice.

I actually think this check is now not necessary anymore, this "file exists" check was created when only str input was allowed.  The try/except error message should be explicit enough. So I propose to remove those two lines entirely. What do you think?

**Comment by yellowcap** [comment]
- Added a test for `vsi_buffer` on non-vsi rasters.
- Removed the additional path check from the `str` based initiation of GDALRasters.
- Squashed the edits commit and added acknowledgements in the commit message. Hope that is ok.

File: django/contrib/gis/gdal/raster/source.py [hunk 14]
```diff
@@ -173,6 +213,25 @@ def _flush(self):
         capi.flush_ds(self._ptr)
 
     @property
+    def vsi_buffer(self):
+        if not self.is_vsi_based:
+            return None
```
Comment (yellowcap) [inline django/contrib/gis/gdal/raster/source.py]: Added test for it [here](https://github.com/django/django/pull/8634/files#diff-469b4a6b934be489c22f3cf20c6e0375R218).
